org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/diagnostics/LSPDiagnosticsToMarkers.java

```diff
@@ -36,28 +31,38 @@
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.NullProgressMonitor;
 import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.jface.text.BadLocationException;
 import org.eclipse.jface.text.IDocument;
 import org.eclipse.jface.text.Position;
 import org.eclipse.jface.text.source.Annotation;
 import org.eclipse.jface.text.source.IAnnotationModel;
 import org.eclipse.jface.text.source.IAnnotationModelExtension;
 import org.eclipse.jface.text.source.ISourceViewer;
+import org.eclipse.lsp4e.ILSPMarker;
 import org.eclipse.lsp4e.LSPEclipseUtils;
 import org.eclipse.lsp4e.LanguageServerPlugin;
+import org.eclipse.lsp4e.MarkerAttributeComputer;
 import org.eclipse.lsp4j.Diagnostic;
 import org.eclipse.lsp4j.PublishDiagnosticsParams;
 import org.eclipse.ui.texteditor.MarkerUtilities;
 
 public class LSPDiagnosticsToMarkers implements Consumer<PublishDiagnosticsParams> {
 
-	public static final String LSP_DIAGNOSTIC = "lspDiagnostic"; //$NON-NLS-1$
-	public static final String LANGUAGE_SERVER_ID = "languageServerId"; //$NON-NLS-1$
 	public static final String LS_DIAGNOSTIC_MARKER_TYPE = "org.eclipse.lsp4e.diagnostic"; //$NON-NLS-1$
 	private final @NonNull String languageServerId;
+	private final @NonNull String markerType;
+	private final @NonNull MarkerAttributeComputer markerAttributeComputer;
 
-	public LSPDiagnosticsToMarkers(@NonNull String serverId) {
+	public LSPDiagnosticsToMarkers(@NonNull String serverId, @Nullable String markerType, @Nullable MarkerAttributeComputer markerAttributeComputer) {
```

Comment:
Yes, that could be done, but to instantiate the LSPDiagnosticsToMarker into the connect method, I would need to use the Supplier patter that we removed before, right? With the current PR, org.eclipse.lsp4e.LanguageServersRegistry.ExtensionLanguageServerDefinition.getDiagnosticHandler(IConfigurationElement) is executed when we instantiate the LanguageClientImpl .

Shall I go back to the version using the supplier and then use LanguageServerWrapper for the initialize?

Reply:
OK.
And could we instantiate and set it all in `LanguageClientImpl.connect` ? Since we have access to the wrapper, we have access to the LanguageServerDefinition, and can try to instantiate the marker handler from here?
Another possibility is to skip the extension part if it's too complex, and allow configuring the marker type and processing extra attributes in the LanguageClientImpl, adding it some default methods `getMarkerTypeForDiagnostic(Diagnostic)` and `processExtraAttributes(markerAttributes)`.
Do you think it would be better or simpler?

Reply:
I hope you like the solution I found, I think it is simple :)